modules/single-port-sg/main.tf

```diff
@@ -40,30 +47,56 @@ variable "udp" {
 }
 
 locals {
-  tcp = "${var.tcp ? 1 : 0}"
-  udp = "${var.udp ? 1 : 0}"
+  tcp       = "${var.tcp ? 1 : 0}"
+  udp       = "${var.udp ? 1 : 0}"
+  by_cidr   = length(var.cidr_blocks) == 0 ? 0 : 1
+  by_src_sg = var.source_security_group_id == "" ? 0 : 1
 }
 
 # ingress rule for tcp, if enabled
-resource "aws_security_group_rule" "tcp_ingress" {
-  count             = local.tcp
+resource "aws_security_group_rule" "tcp_ingress_cidr" {
+  count             = local.tcp * local.by_cidr
   type              = "ingress"
   description       = "${var.description} (tcp)"
   from_port         = var.port
   to_port           = var.port
   protocol          = "tcp"
-  cidr_blocks       = var.cidr_blocks
   security_group_id = var.security_group_id
+  cidr_blocks       = var.cidr_blocks
 }
 
 # ingress rule for udp, if enabled
-resource "aws_security_group_rule" "udp_ingress" {
-  count             = local.udp
+resource "aws_security_group_rule" "udp_ingress_cidr" {
+  count             = local.udp * local.by_cidr
   type              = "ingress"
   description       = "${var.description} (udp)"
   from_port         = var.port
   to_port           = var.port
   protocol          = "udp"
-  cidr_blocks       = var.cidr_blocks
   security_group_id = var.security_group_id
+  cidr_blocks       = var.cidr_blocks
+}
+
+# ingress rule for tcp, if enabled
+resource "aws_security_group_rule" "tcp_ingress_src_sg" {
+  count                    = local.tcp * local.by_src_sg
```

Comment:
Define the logic/math as some local value we refer to here. For example, this is `enable_tcp_by_src_sg` (or `enable_tcp_by_sg`) and then there is `enable_tcp_by_cidr`

Reply:
Hi, I had a discussion in Gitlab. TL;DR, the whole function does not work well right now.

If the source security group does not exist yet, TF fails to calculate the numbers. Targeting the source SG first makes this work, but not ideal. So we need new design, for example, a different module to support source security group (and not calculating).